src/main/java/org/apache/sysds/runtime/matrix/data/MatrixBlock.java

```diff
@@ -643,7 +643,36 @@ public void quickSetValue(int r, int c, double v)
 				nonZeros--;
 		}
 	}
-	
+
+	/*
+		Thread save set.
+		Blocks need to be allocated and in case of MCSR sparse all rows that are going to be accessed need to be allocated
+		aswell.
+	 */
+	public void quickSetValueThreadSafe(int r, int c, double v){
+		if(sparse){
+			if(!(sparseBlock instanceof SparseBlockMCSR))
+				throw new RuntimeException("Only MCSR Blocks are supported for Multithreaded sparse set.");
+			synchronized (sparseBlock.get(r)){
+				sparseBlock.set(r,c,v);
+			}
+		}else{
+			denseBlock.set(r,c,v);
+		}
```

Comment:
Since denseblocks are just a collection of arrays and at the moment of writing to the block there should be nothing that is reading we can write concurrently without any need of synchronization, on the other hand the sparse blocks are only row independent so we need to sync over rows.